### PR TITLE
#907, #891, #821: Correct and document the release procedure

### LINKED ISSUE #891: Support the old URL to the PDF of the reference manual
I am responsible for creating many links around the internet over the years to the URL `https://github.com/dafny-lang/dafny/blob/master/Docs/DafnyRef/out/DafnyRef.pdf`, which used to host a PDF of the reference manual. Is there any chance we make the build system put a copy of the new reference manual in PDF format at that location?

### LINKED ISSUE #907: The java compilation of a very simple code snippet fails as the compiler cannot find the symbol dafny.Type
## Issue
When I execute the command `dafny /compile:1 /compileTarget:java /spillTargetCode:1 test.dfy` on the following code snippet 
```
function method identity(n:nat):nat
{
    n
}
```
I get the following outputs
### Terminal
```
Dafny 3.0.0.20820

Dafny program verifier finished with 0 verified, 0 errors
File test/test.java contains the partially compiled program
Additional code written to test/_System/nat.java
Additional code written to test/_System/__default.java
/home/roberto/test/_System/nat.java:10: error: cannot find symbol
  private static final dafny.Type<java.math.BigInteger> _TYPE = dafny.Type.referenceWithInitializer(java.math.BigInteger.class, () -> java.math.BigInteger.ZERO);
                            ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/_System/nat.java:11: error: cannot find symbol
  public static dafny.Type<java.math.BigInteger> _type() {
                     ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/_System/__default.java:13: error: cannot find symbol
  private static final dafny.Type<__default> _TYPE = dafny.Type.referenceWithInitializer(__default.class, () -> (__default) null);
                            ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/_System/__default.java:14: error: cannot find symbol
  public static dafny.Type<__default> _type() {
                     ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/dafny/Tuple0.java:8: error: cannot find symbol
  private static final dafny.Type<Tuple0> _TYPE = dafny.Type.referenceWithInitializer(Tuple0.class, () -> Default());
                            ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/dafny/Tuple0.java:9: error: cannot find symbol
  public static dafny.Type<Tuple0> _type() {
                     ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/_System/nat.java:10: error: cannot find symbol
  private static final dafny.Type<java.math.BigInteger> _TYPE = dafny.Type.referenceWithInitializer(java.math.BigInteger.class, () -> java.math.BigInteger.ZERO);
                                                                     ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/_System/nat.java:12: error: cannot find symbol
    return (dafny.Type<java.math.BigInteger>) (dafny.Type<?>) _TYPE;
                 ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/_System/nat.java:12: error: cannot find symbol
    return (dafny.Type<java.math.BigInteger>) (dafny.Type<?>) _TYPE;
                                                    ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/_System/__default.java:13: error: cannot find symbol
  private static final dafny.Type<__default> _TYPE = dafny.Type.referenceWithInitializer(__default.class, () -> (__default) null);
                                                          ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/_System/__default.java:15: error: cannot find symbol
    return (dafny.Type<__default>) (dafny.Type<?>) _TYPE;
                 ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/_System/__default.java:15: error: cannot find symbol
    return (dafny.Type<__default>) (dafny.Type<?>) _TYPE;
                                         ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/dafny/Tuple0.java:8: error: cannot find symbol
  private static final dafny.Type<Tuple0> _TYPE = dafny.Type.referenceWithInitializer(Tuple0.class, () -> Default());
                                                       ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/dafny/Tuple0.java:10: error: cannot find symbol
    return (dafny.Type<Tuple0>) (dafny.Type<?>) _TYPE;
                 ^
  symbol:   class Type
  location: package dafny
/home/roberto/test/dafny/Tuple0.java:10: error: cannot find symbol
    return (dafny.Type<Tuple0>) (dafny.Type<?>) _TYPE;
                                      ^
  symbol:   class Type
  location: package dafny
15 errors
[ERROR] FATAL UNHANDLED EXCEPTION: System.Exception: Error while compiling Java files. Process exited with exit code 1
  at Microsoft.Dafny.JavaCompiler.CompileTargetProgram (System.String dafnyProgramName, System.String targetProgramText, System.String callToMain, System.String targetFilename, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, System.Boolean hasMain, System.Boolean runAfterCompile, System.IO.TextWriter outputWriter, System.Object& compilationResult) [0x001d8] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.DafnyDriver.CompileDafnyProgram (Microsoft.Dafny.Program dafnyProgram, System.String dafnyProgramName, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, System.Boolean invokeCompiler, System.IO.TextWriter outputWriter) [0x0034e] in <6fd7bea9fd0c48cc9a4e825f7a106a9b>:0 
  at Microsoft.Dafny.DafnyDriver.Compile (System.String fileName, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.Program dafnyProgram, Microsoft.Boogie.PipelineOutcome oc, System.Collections.Generic.Dictionary`2[TKey,TValue] statss, System.Boolean verified) [0x0005f] in <6fd7bea9fd0c48cc9a4e825f7a106a9b>:0 
  at Microsoft.Dafny.DafnyDriver.ProcessFiles (System.Collections.Generic.IList`1[T] dafnyFiles, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.ErrorReporter reporter, System.Boolean lookForSnapshots, System.String programId) [0x0023a] in <6fd7bea9fd0c48cc9a4e825f7a106a9b>:0 
  at Microsoft.Dafny.DafnyDriver.ThreadMain (System.String[] args) [0x0004b] in <6fd7bea9fd0c48cc9a4e825f7a106a9b>:0 
  at Microsoft.Dafny.DafnyDriver+<>c__DisplayClass1_0.<Main>b__0 () [0x00001] in <6fd7bea9fd0c48cc9a4e825f7a106a9b>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00014] in <9f0df102fe6e4cfea29d2e46f585d8a5>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <9f0df102fe6e4cfea29d2e46f585d8a5>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <9f0df102fe6e4cfea29d2e46f585d8a5>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <9f0df102fe6e4cfea29d2e46f585d8a5>:0 
  at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <9f0df102fe6e4cfea29d2e46f585d8a5>:0 
```
### Java Files
image

## Versions Affected
- 3.0.0 pre-release 0a
- 3.0.0 pre-release 1

on Ubuntu 18.04

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #907, #891, #821.

### CODE REVIEW

File: Test/git-issues/git-issue-590.dfy [hunk 1]
```diff
@@ -1,3 +1,3 @@
-// RUN: %dafny /compile:3 x.cs > "%t"
+// dafny should emit exit value 1
+// RUN: %dafny /compile:3 x.cs > "%t" || echo ERROR EXIT >> "%t"
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-590.dfy:2]: Does `||` work on Windows?

File: RELEASE.txt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+4) Create a new release: On 'https://github.com/dafny-lang/dafny/releases'
+   click 'Draft a new release'
+
+5) Fill in a tag ('v' followed by release name), the release name,
+the release notes.
+
+6) Add all the .zip files created by step 3 in the Package folder.
+
+7) Commit and push the new version of DafnyRef.pdf (this version is also
+   included in the releasae .zip files)
```
Comment (RustanLeino) [inline RELEASE.txt:21]:
```suggestion
   included in the release .zip files)
```

File: RELEASE.txt [hunk 3]
```diff
@@ -0,0 +1,25 @@
+
+This is the procedure for generating a new Dafny release.
+
```
Comment (RustanLeino) [inline RELEASE.txt:3]: * Another step is to update the version number in dafny/Source/version.cs to be $VER. This should be done before step 3. The updated file should be pushed.
* Around step 5-6, one should also write some release notes.
* Github will ask for a name of the release. Use “Dafny $V”, where $V are the first three components of the 4-component $VER.

File: RELEASE.txt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+4) Create a new release: On 'https://github.com/dafny-lang/dafny/releases'
+   click 'Draft a new release'
+
+5) Fill in a tag ('v' followed by release name), the release name,
+the release notes.
+
+6) Add all the .zip files created by step 3 in the Package folder.
+
+7) Commit and push the new version of DafnyRef.pdf (this version is also
```
Comment (RustanLeino) [inline RELEASE.txt:20]: From where does one obtain DafnyRef.pdf?

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: Test/git-issues/git-issue-590.dfy (hunk 1 above)
Comment (davidcok) [reply Test/git-issues/git-issue-590.dfy:2]: Aargh no it won't when we combine with the expanded CI tests. I'll change countVerificationErrors to suppress command-line errors as well -- and hope Robin can finish the test infrastructure soon

File: RELEASE.txt (hunk 3 above)
Comment (davidcok) [reply RELEASE.txt:3]: OK

File: RELEASE.txt (hunk 4 above)
Comment (davidcok) [reply RELEASE.txt:20]: Enhanced the instructions -- it is already in the workspace -- just needs to be committed to git.

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**